File: web/src/lib/api/schema/stats.ts
/**
 * Schema for the `GET /api/stats` response.
 *
 * Mirrors the `StatsResponse` DTO emitted by yog-api: protocol-wide
 * aggregate statistics backing the Overview page (but client-agnostic —
 * the endpoint ships raw counters, the UI derives the coverage label).
 *
 * Notes on the fields:
 *   - USD aggregates are `BigDecimal` strings (like `PoolResponse`) to
 *     preserve the full precision the SQL produces; `null` when nothing
 *     is priceable / no activity in the window. The UI formats them.
 *   - `poolsPriced` is the coverage numerator (pools that contributed to
 *     `totalTvlUsd`); `poolsObserved` is the denominator. Both are JSON
 *     numbers (BIGINT counts, well within the JS safe range here).
 *   - `poolsDiscovered24h` counts pools first seen in the last 24h.
 */

import * as z from "zod";

import { BigDecimal } from "./shared";

/** Schema for the full protocol-wide stats payload. */
export const StatsSchema = z.object({
  // Summed current TVL across priceable pools; null when none is priceable.
  totalTvlUsd: BigDecimal.nullable(),
  // How many pools contributed to totalTvlUsd (coverage numerator).
  poolsPriced: z.number().int().nonnegative(),
  // Summed realized volume / trading fee over the last 24h (trade-time valued).
  volume24hUsd: BigDecimal.nullable(),
  fees24hUsd: BigDecimal.nullable(),
  // Every pool ever observed (coverage denominator).
  poolsObserved: z.number().int().nonnegative(),
  // Pools first seen in the last 24h (the discovery pulse).
  poolsDiscovered24h: z.number().int().nonnegative(),
});

/** Validated protocol-wide stats payload. */
export type StatsResponse = z.infer<typeof StatsSchema>;

File: crates/api/src/http/dto/response/stats.rs
//! Response DTO for `GET /api/stats`.
//!
//! Protocol-wide aggregate statistics. Kept separate from the domain types
//! like every other `*Response` — the domain model never leaks into the wire
//! shape. Decimals serialize the same way as `PoolResponse` (rust_decimal's
//! native representation).
//!
//! The endpoint is intentionally client-agnostic: it ships raw counters
//! (`poolsPriced` + `poolsObserved`) rather than a formatted "N / M priced"
//! coverage string — the presentation layer derives that. Same for the LP fee
//! share or effective rate: not the API's job here.

use rust_decimal::Decimal;
use serde::Serialize;

use crate::application::StatsAggregate;

/// Protocol-wide statistics payload.
///
/// Three USD aggregates plus two pool counters:
///   - `total_tvl_usd` + `pools_priced`: summed current TVL and how many pools
///     it covers (the coverage numerator; denominator is `pools_observed`).
///   - `volume_24h_usd` / `fees_24h_usd`: summed realized volume and trading
///     fee over the last 24h, trade-time valued.
///   - `pools_observed`: every pool ever seen. `pools_discovered_24h`: those
///     first seen in the last 24h (the discovery pulse).
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct StatsResponse {
    total_tvl_usd: Option<Decimal>,
    pools_priced: i64,
    volume_24h_usd: Option<Decimal>,
    fees_24h_usd: Option<Decimal>,
    pools_observed: i64,
    pools_discovered_24h: i64,
}

impl From<StatsAggregate> for StatsResponse {
    fn from(agg: StatsAggregate) -> Self {
        Self {
            total_tvl_usd: agg.analytics.total_tvl_usd,
            pools_priced: agg.analytics.pools_priced,
            volume_24h_usd: agg.analytics.volume_24h_usd,
            fees_24h_usd: agg.analytics.fees_24h_usd,
            pools_observed: agg.counts.observed,
            pools_discovered_24h: agg.counts.discovered_24h,
        }
    }
}
